flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/history/HistoryServerTest.java

```diff
@@ -117,45 +124,145 @@ public void testHistoryServerIntegration() throws Exception {
+				if (event.getType() == HistoryServerArchiveFetcher.ArchiveEventType.CREATED) {
+					numExpectedArchivedJobs.countDown();
+				}
+			});
+
+		try {
+			hs.start();
+			String baseUrl = "http://localhost:" + hs.getWebPort();
+			numExpectedArchivedJobs.await(10L, TimeUnit.SECONDS);
+
+			ObjectMapper mapper = new ObjectMapper();
+			Assert.assertEquals(numJobs + numLegacyJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+
+		} finally {
+			hs.stop();
+		}
+	}
+
+	@Test
+	public void testCleanExpiredJob() throws Exception {
+		runArchiveExpirationTest(true);
+	}
+
+	@Test
+	public void testRemainExpiredJob() throws Exception {
+		runArchiveExpirationTest(false);
+	}
+
+	private void runArchiveExpirationTest(boolean cleanupExpiredJobs) throws Exception {
+		int numExpiredJobs = cleanupExpiredJobs ? 1 : 0;
+		int numJobs = 3;
+		List<String> jobIDs = new ArrayList<>();
+		for (int x = 0; x < numJobs; x++) {
+			jobIDs.add(runJob());
 		}
+		waitForArchivesCreation(numJobs);
+
+		CountDownLatch numExpectedArchivedJobs = new CountDownLatch(numJobs);
+		CountDownLatch numExpectedExpiredJobs = new CountDownLatch(numExpiredJobs);
+
+		Configuration historyServerConfig = createTestConfiguration(cleanupExpiredJobs);
+
+		HistoryServer hs =
+			new HistoryServer(
+				historyServerConfig,
+				(event) -> {
+					switch (event.getType()){
+						case CREATED:
+							numExpectedArchivedJobs.countDown();
+							break;
+						case DELETED:
+							numExpectedExpiredJobs.countDown();
+							break;
+					}
+				});
 
-		HistoryServer hs = new HistoryServer(historyServerConfig, numExpectedArchivedJobs);
 		try {
 			hs.start();
+			// check that expected count of archives exists
 			String baseUrl = "http://localhost:" + hs.getWebPort();
 			numExpectedArchivedJobs.await(10L, TimeUnit.SECONDS);
 
 			ObjectMapper mapper = new ObjectMapper();
-			String response = getFromHTTP(baseUrl + JobsOverviewHeaders.URL);
-			MultipleJobsDetails overview = mapper.readValue(response, MultipleJobsDetails.class);
+			Assert.assertEquals(numJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+
+			String jobIDToDelete = jobIDs.get(0);
+			verifyArchiveFilesExistence(jobIDToDelete, true);
+
+			// delete one archive from jm
+			URI resolve = jmDirectory.toURI().resolve(jobIDToDelete);
+			new File(resolve).delete();
+
+			numExpectedExpiredJobs.await(10L, TimeUnit.SECONDS);
+
+			// check that archive presence in hs
+			Assert.assertEquals(numJobs - numExpiredJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+			verifyArchiveFilesExistence(jobIDToDelete, !cleanupExpiredJobs);
 
-			Assert.assertEquals(numJobs + 1, overview.getJobs().size());
 		} finally {
 			hs.stop();
 		}
 	}
 
-	private static void runJob() throws Exception {
+	private void verifyArchiveFilesExistence(String jobID, boolean expectArchiveFilesExist) {
+		List<File> archiveFiles = Arrays.asList(
+			new File(hsDirectory.toURI().resolve(String.format("overviews/%s.json", jobID))),
```

Comment:
`hsDirectory.toPath()` is more common in Flink I believe; additionally, once you have a Path you could work agains the nio Files API, which tends to provide better error messages in case of errors.

Reply:
Changed to list of `Path`s, but finally whole method was deleted because you didn't like that internal files hierarchy is tested.